apps/barista-design-system/src/shared/services/scroll-spy.service.ts

```diff
@@ -14,212 +14,123 @@
-    if (scrollTop + 1 >= maxScrollTop) {
-      activeItem = this._spiedElements[0];
-    } else {
-      activeItem = this._spiedElements.find((spiedElem) => {
-        if (spiedElem.top <= scrollTop) {
-          return true;
-        }
-        return false;
-      });
-    }
-
-    this.activeScrollItem.next(activeItem || null);
+  /** Calculates top bounding properties for an element array */
+  calcTopValues(elements: Element[]): void {
+    this.headlineElements = [];
+    this.headlineElements = elements
+      .map((element) => ({
+        top:
+          this._getScrollTop() +
+          element.getBoundingClientRect().top -
+          this.OFFSET,
+        element,
+      }))
+      .sort((a, b) => compareValues(a.top, b.top, 'asc'));
   }
-}
-
-@Injectable()
-export class BaScrollSpyService {
-  private _spiedElementGroups: BaScrollSpiedElementGroup[] = [];
-  private _onStopListening = new Subject<void>();
-  private _lastContentHeight: number;
-  private _lastMaxScrollTop: number;
-  private _topOffset = 66;
-
-  // tslint:disable-next-line: no-any
-  constructor(
-    private _platform: Platform,
-    @Inject(DOCUMENT) private _document: any,
-  ) {}
 
-  /*
-   * Start tracking a group of elements and emitting active elements; i.e. elements that are
-   * currently visible in the viewport. If there was no other group being spied, start listening for
-   * `resize` and `scroll` events.
+  /**
+   * Evaluates by comparing the users scroll position with the element top property and return the id of an element
+   * that should be highlighted
    */
-  spyOn(elements: Element[]): BaScrollSpyInfo | null {
-    if (this._platform.isBrowser && !this._spiedElementGroups.length) {
-      fromEvent(window, 'resize')
-        .pipe(auditTime(300), takeUntil(this._onStopListening))
-        .subscribe(() => {
-          this._onResize();
-        });
-      fromEvent(window, 'scroll')
-        .pipe(auditTime(10), takeUntil(this._onStopListening))
-        .subscribe(() => {
-          this._onScroll();
-        });
-      this._onResize();
-
+  findActiveItem(): string | null {
+    // The element id of the item to be highlighted
+    let activeItem: string | null = null;
+    this._zone.runOutsideAngular(() => {
       const scrollTop = this._getScrollTop();
-      const maxScrollTop = this._lastMaxScrollTop;
-
-      const spiedGroup = new BaScrollSpiedElementGroup(elements);
-      spiedGroup.calibrate(scrollTop, this._topOffset);
-      spiedGroup.onScroll(scrollTop, maxScrollTop);
-
-      this._spiedElementGroups.push(spiedGroup);
-
-      return {
-        active: spiedGroup.activeScrollItem
-          .asObservable()
-          .pipe(distinctUntilChanged()),
-        unspy: () => {
-          this._unspy(spiedGroup);
-        },
-      };
-    }
+      const contentHeight = this._getContentHeight();
+      const viewportHeight = this._getViewportHeight();
+      for (let i = 0; i < this.headlineElements.length; i++) {
```

Comment:
Maybe a while loop is better suited here since you don't have to continue the loop once you have your active item.